### ISSUE TITLE
[Feature]: Pinning versions in CPM should allow nuspec dependencies to not be added

### ISSUE THREAD

**Issue by namerril** [NuGet/Home#12128]
### NuGet Product(s) Involved

NuGet.exe

### The Elevator Pitch

A NuGet package should only declare dependencies on packages that they actually use.  However, if I have enabled pinning for my repository, then it is impossible for my package to not directly depend on a transitive dependency, even if that dependency is spurious.

If this situation occurs, then the only option I have at the moment is to remove pinning for the project.  I'd prefer a solution where I can indicate that a particular PackageVersion that is pinned should be private (e.g. PrivateAssets="all"), so that the output nuspec does not contain it as a dependency.

**Comment by jeffkl** [comment]
@namerril Let me see if I understand the feature ask correctly.  You have a reference to `PackageA` which depends on `PackageB` `1.0.0`.  But you want to pin `PackageB` to version `2.0.0` so you enable central package management pinning and supply a `<PackageVersion Include="PackageB" Version="2.0.0" />`.

Now `PackageB` shows up as a direct dependency on your package since NuGet knows that you used a different version than what `PackageA` would have pulled in transitively.  But in your case, `PackageB` is "private" in that you don't need it to flow to your downstream dependencies so you want to put `PrivateAssets="All"` on the package reference but you can't because its pinned and not declared as a top-level dependency.   Essentially, metadata from `<PackageVersion />` does not flow to `<PackageReference />` items?

**Comment by nathanmerrill** [comment]
I'm not sure if I would say broadly say "metadata from PackageVersion does not flow".  There's other metadata on PackageVersion that could be useful.  However, pretty much everything else is accurate.

Adding `PrivateAssets="All"` on a PackageVersion is simply an idea of how this could work.  I think it makes sense (because it matches known behaviors).  However, the real end goal here is "Allow my project to pin a version without adding it to the nuspec somehow".

**Comment by jeffkl** [comment]
Okay so the only way to currently do this is to pin it yourself by declaring PackageB as a top-level dependency:

```xml
<PackageReference Include="PackageB" PrivateAssets="All" />
```

We don't currently flow `PrivateAssets` from `PackageVersion` to `PackageReference` because that would apply to every project in the tree.  You'd need to do something fancy like 

```xml
<PackageVersion Include="PackageB" Version="2.0.0" />
<PackageVersion Update="PackageB" PrivateAssets="All" Condition="'$(ThisProjectWantsPackageBPrivate)' == 'true'" />
```

Adding the `<PackageReference />` item to the project in my opinion is more explicit.  It obviously doesn't scale well if you want to pin PackageB with PrivateAssets=All in your whole tree.

**Comment by nathanmerrill** [comment]
For situations where I want this feature, I typically want it for the whole tree.  E.g. there's a security issue with a particular version, and I don't want to build against it.  If I really wanted it it for a single project, I'd just add the PackageVersion Update to the project itself.

**Comment by nathanmerrill** [comment]
Actually, I'm unsure if PrivateAssets="All" makes sense here.  The primary argument against it is existing PackageReferences in other projects.  For example, consider:

```xml
<PackageReference Include="PackageB" />
<PackageVersion Include="PackageB" Version="2.0.0" PrivateAssets="All" />
```

This is confusing.  Is PackageB private in this project?  Ideally, the answer should be no:  It would only make it private for projects that didn't already declare the PackageReference.

Therefore, I think a separate property for a pinned version like `PinAsPrivate="true"` might make more sense

**Comment by dfederm** [comment]
> "Allow my project to pin a version without adding it to the nuspec somehow"

Doesn't this mean that you compile against a pinned version (let's say version 2) but as there's no entry in the nuspec, any consumer may end up with some other version, for example version 1?

At best that seems to push the requirement of pinning that package version downstream (why would you need to pin but consumers wouldn't?), and at worst just lead to hidden runtime errors as the consumer wouldn't get the "downgrade detected" warnings they should be getting in that case.

**Comment by nathanmerrill** [comment]
It's a transitive dependency, not a direct dependency.  My direct dependency A was originally compiled against version 1, so no runtime issues should occur by forcing the version to 2 in my local build (assuming 1 -> 2 isn't a breaking change)

**Comment by dfederm** [comment]
Transitive dependencies are still provided to the compiler, so you are able to still directly use them even if you aren't directly PackageReferencing them. So there would need to be extra compiler involvement to validate "ok you aren't actually using this indirect dependency". But in that case, why bother pinning your indirect dependency to a higher version anyway? If it's just for CG / security, I'd argue that for the same exact reasons outlined above, you're not "actually" using the dependency anyway so pinning is just checking an arbitrary checkbox which has no real world impact.

Also, the pushing of the pinning requirement to your consumers is still a problem here.

**Comment by nathanmerrill** [comment]
I don't think we need the compiler to validate this.  It's possible to force invalid dependency trees using other methods (PrivateAssets, for example).  If a pinned version is not actually transitive, and should be declared as a direct dependency, then that's an issue that the package already had, and this makes that issue worse.  I wouldn't be against a compiler check, I just don't think it's strictly necessary, but that choice isn't up to me :)

For our use case, we ensure that packages only declare their actual direct dependencies, and no more.  We also want to not use vulnerable packages using pinning.  The only way I can see to fulfill both of these requirements is to add pins to upgrade vulnerable versions, but also not to add pinned versions to nuspec.

I do see value in patching versions for downstream consumers.  But that should be a choice that package producer makes.
